File: Igneel.Physics/PhysicManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Igneel.Collections;
using System.IO;


namespace Igneel.Physics
{
    public abstract class PhysicManager:ResourceAllocator
    {
        internal ReadOnlyDictionary<string, Physic> scenes = new ReadOnlyDictionary<string, Physic>();
        internal ReadOnlyDictionary<string, TriangleMesh> meshes = new ReadOnlyDictionary<string, TriangleMesh>();

        static PhysicManager _instance;

        public PhysicManager()
        {
            Service.Set<PhysicManager>(this);
            _instance = this;
        }

        public static PhysicManager Sigleton { get { return _instance; } }

        public ReadOnlyDictionary<string, TriangleMesh> Meshes { get { return meshes; } }

        public ReadOnlyDictionary<string, Physic> Scenes { get { return scenes; } }

        public Physic CreatePhysic(PhysicDesc desc)
        {
            var scene = _CreatePhysic(desc);
            scenes.Add(scene.Name, scene);
            return scene;
        }

        public TriangleMesh CreateTriangleMesh(TriangleMeshDesc desc)
        {
            var mesh = _CreateTriangleMesh(desc);
            meshes.Add(mesh.Name, mesh);

            var srv = Service.Get<INotificationService>();
            if (srv != null)
                srv.OnObjectCreated(mesh);

            return mesh;
        }

        public TriangleMesh CreateTriangleMeshFromFile(string filename)
        {
            var mesh = _CreateTriangleMeshFromFile(filename);
            mesh.Name = "Mesh" + meshes.Count;
            meshes.Add(mesh.Name, mesh);

            var srv = Service.Get<INotificationService>();
            if (srv != null)
                srv.OnObjectCreated(mesh);
            return mesh;
        }

        public TriangleMesh CreateTriangleMeshFromStream(Stream stream)
        {
            var mesh = _CreateTriangleMeshFromStream(stream);
            meshes.Add(mesh.Name, mesh);
            var srv = Service.Get<INotificationService>();
            if (srv != null)
                srv.OnObjectCreated(mesh);
            return mesh;
        }

        public abstract CharacterControllerManager CreateControllerManager();

        protected abstract Physic _CreatePhysic(PhysicDesc desc);

        protected abstract TriangleMesh _CreateTriangleMesh(TriangleMeshDesc desc);

        protected abstract TriangleMesh _CreateTriangleMeshFromFile(string filename);

        protected abstract TriangleMesh _CreateTriangleMeshFromStream(Stream stream);

        protected override void OnDispose(bool disposing)
        {          
            foreach (var m in meshes.ToArray())
            {
                m.Value.Dispose();
            }
            foreach (var s in scenes.ToArray())
            {
                s.Value.Dispose();
            }
            base.OnDispose(disposing);
        }
    }
}

File: Engine/Igneel.PhysX/PhysicManager.cpp
#include "Stdafx.h"
#include "PhysicManager.h"
#include "OutputStream.h"
#include "XScene.h"
#include "Stream.h"
#include "XTriangleMesh.h"
#include "XControllerManager.h"

namespace Igneel { namespace PhysX {

	PXPhysicManager::PXPhysicManager()
	{
		NxSDKCreateError error;			
		NxPhysicsSDKDesc desc = NxPhysicsSDKDesc();			
		MyUserOutputStream * m_outputStream = new MyUserOutputStream();

		//sdk = NxCreatePhysicsSDK(NX_PHYSICS_SDK_VERSION);
		sdk =  NxCreatePhysicsSDK(NX_PHYSICS_SDK_VERSION, NULL, m_outputStream, desc, &error);		
		if(!sdk)	
		{
			delete m_outputStream;
			throw gcnew InvalidOperationException(L"Invalid PhysX");
		}

		cooking = NxGetCookingLib(NX_PHYSICS_SDK_VERSION);
		cooking->NxInitCooking();

		GC::AddMemoryPressure(sizeof(NxPhysicsSDK));		
		GC::AddMemoryPressure(sizeof(NxCookingInterface));		
	}

	void PXPhysicManager::OnDispose(bool disposing)
	{
		__super::OnDispose(disposing);		
		if(cooking)
		{
			cooking->NxCloseCooking();
			cooking = NULL; 
		}
		if(sdk)
		{			
			NxReleasePhysicsSDK(sdk);
			sdk = NULL;			
		}		
	}

	Physic^ PXPhysicManager::_CreatePhysic(PhysicDesc^ desc)
	{
		return gcnew XScene(desc);
	}

	TriangleMesh^ PXPhysicManager::_CreateTriangleMesh(TriangleMeshDesc^ desc)
	{
		if(!cooking)
			throw gcnew NullReferenceException(L"cooking sdk null");
		
		NxTriangleMeshDesc md = NxTriangleMeshDesc();
		md.flags = static_cast<NxU32>( desc->Flags );
		md.numTriangles = desc->NumTriangles;
		md.numVertices = desc->NumVertices;
		md.points = desc->Points.ToPointer();
		md.pointStrideBytes = desc->PointStrideBytes;
		md.triangles = desc->Triangles.ToPointer();
		md.triangleStrideBytes = desc->TriangleStrideBytes;		

		if(!md.isValid())
			throw gcnew InvalidOperationException(L"Invalid Mesh");

		MemoryWriteBuffer writer;
		cooking->NxCookTriangleMesh(md, writer);		

		MemoryReadBuffer mb(writer.data);
		NxTriangleMesh* tmesh = sdk->createTriangleMesh(mb);

		auto triangleMesh = gcnew XTriangleMesh(tmesh);
		triangleMesh->Name = desc->Name;
		return triangleMesh;
	}

	TriangleMesh^ PXPhysicManager::_CreateTriangleMeshFromFile(String^ filename)
	{
		IntPtr pter  = Marshal::StringToHGlobalAnsi(filename);

		UserStream us((const char*)pter.ToPointer(),true);

		auto tmesh = sdk->createTriangleMesh(us);
		if(!tmesh)
			throw gcnew InvalidOperationException(L"Invalid Data");

		Marshal::FreeHGlobal(pter);

		auto triangleMesh = gcnew XTriangleMesh(tmesh);		
		return triangleMesh;
	}

	TriangleMesh^ PXPhysicManager::_CreateTriangleMeshFromStream(Stream^ stream)
	{
		BinaryReader^ reader = gcnew BinaryReader(stream);
		BinaryWriter^ wrtter = gcnew BinaryWriter(stream);

		WrapperStream ws(reader, wrtter);

		auto tmesh = sdk->createTriangleMesh(ws);
		if(!tmesh)
			throw gcnew InvalidOperationException(L"Invalid Data");

		stream->Close();

		return gcnew XTriangleMesh(tmesh);
	}

	CharacterControllerManager^ PXPhysicManager::CreateControllerManager()
	{
		return gcnew XControllerManager();
	}
}}